### ISSUE TITLE
Question about single-object training: how are clicks generated for binary-class tasks?

### ISSUE THREAD

**Issue by 772240177** [ywyue/AGILE3D#17]
Dear Yue! thank you for your great work!

I'm trying to understand how the single-object training is handled in your code. I noticed there isn't a dedicated training script for the single-object mode, and I couldn't find details in the documentation. So I have a few questions:

If I have a dataset with N classes, does the single-object training randomly select one class as the target for each sample?

For binary segmentation tasks (e.g., foreground/background or cat/dog), does the training sample clicks for both classes randomly, or does it only generate clicks for one class (e.g., only foreground)?

I’d really appreciate it if you could clarify how click sampling is done in the single-object training setting.

Thanks in advance!

**Comment by ywyue** [comment]
hi @772240177, thanks for your interest and kind words!

Yes, we only train a single model in multi-object setup and evaluate it in both multi-object and single-object setup.

To make the code work for single-object training, you can simply set `num_obj` [here](https://github.com/ywyue/AGILE3D/blob/45ab50c2f3bbe45f4027c7966f6eb52d8ef23f55/engine.py#L65C13-L65C20) as 1. This is the minimum modification required for the single-object training code, but it is not optimal. Ideally, we should use `datasets/InterSingleObj3DSegDataset.py` for single-object training.

> If I have a dataset with N classes, does the single-object training randomly select one class as the target for each sample?

We assume you have instance mask annotation not semantic mask. If one scene in your dataset has N object masks, then the above solution will randomly select one object as the target object.

> For binary segmentation tasks (e.g., foreground/background or cat/dog), does the training sample clicks for both classes randomly, or does it only generate clicks for one class (e.g., only foreground)?

We use iterative click sampling not random click sampling. In more detail, we sample the first click from the center of the target object and then sample the next click from the biggest error region, and this click can be at both the target object or background.

**Comment by 772240177** [comment]
Thank you for your reply! I checked the code you pointed out, and I saw that you set num_obj to 1 for single-object training, specifically using obj_idxs = valid_obj_idxs[torch.randperm(max_num_obj)[:1]]. This means one instance is randomly selected, and in the case of binary classification, this setup allows both classes to be involved in training.
